data/shaders/opengl/rayleigh.glsl

```diff
@@ -0,0 +1,237 @@
+}
+
+// given a and b segments, return c = a \ b
+vec4 segmentSubtraction(const in vec2 a, const in vec2 b)
+{
+	vec2 as = sortAscending(a);
+	vec2 bs = sortAscending(b);
+
+	// b could be inside a, leaving segments at both sides
+	vec4 c;
+	c.x = a.x;
+	c.w = a.y;
+
+	c.y = min(a.y, max(a.x, b.x));
+	c.z = max(a.x, min(a.y, b.y));
+	return c;
+}
+
+vec3 computeIncidentLight(const in vec3 sunDirection, const in vec3 dir, const in vec3 center, const in vec2 atmosDist, const in vec4 diffuse, const in float uneclipsed)
+{
+	vec3 betaR = vec3(3.8e-6f, 13.5e-6f, 33.1e-6f);
+	vec3 betaM = vec3(21e-6f);
+
+	float atmosMin = atmosDist.x * geosphereRadius;
+	float atmosMax = atmosDist.y * geosphereRadius;
+
+	// solve Cylinder entry/exit dist
+	vec2 cylinder_intersect = rayCylinderIntersect(dir, center, sunDirection, geosphereRadius);
+	bool hasIntersect = cylinder_intersect.x != 0 || cylinder_intersect.y != 0;
+
+	vec3 cylinder_near = center - dir * cylinder_intersect.x;
+	vec3 cylinder_far  = center - dir * cylinder_intersect.y;
+
+	// test if ray passes through shadow
+	float a = dot(cylinder_near, sunDirection);
+	float b = dot(cylinder_far , sunDirection);
+	bool intersectsShadow = hasIntersect && (a > 0.f || b > 0.f);
+
+	vec2 ground_intersect = raySphereIntersect(center, dir, geosphereRadius);
+	bool shadowVisible = intersectsShadow && ground_intersect.x == 0.f;
+
+	/*
+	* We have three options:
+	* 1) Ray does not intersect shadow
+	*    Do nothing
+	* 2) Ray intersects shadow, starts inside
+	*    (cylinder_intersect.y, tmax)
+	* 3) Ray intersects shadow, starts outside
+	*    (tmin, cylinder_intersect.x) + (cylinder_intersect.y, tmax)
+	*/
+
+	vec2 atmosphere_intersect = raySphereIntersect(center, dir, geosphereRadius * geosphereAtmosTopRad);
+	vec4 atmosphere_minus_shadow = shadowVisible ? segmentSubtraction(atmosphere_intersect, cylinder_intersect) : vec4(atmosphere_intersect.x, atmosphere_intersect.y, atmosphere_intersect.y, atmosphere_intersect.y);
+
+	if (ground_intersect.x > 0.f) {
+		atmosphere_minus_shadow.yzw = min(atmosphere_minus_shadow.yzw, ground_intersect.x);
+	}
+	if (ground_intersect.y < 0.f) {
+		atmosphere_minus_shadow.xyz = max(atmosphere_minus_shadow.xyz, ground_intersect.y);
+	}
+
+	vec3 sumR = vec3(0.f);
+	vec3 sumM = vec3(0.f);
+	vec2 opticalDepth = vec2(0.f);
+
+	processRay(sumR, sumM, opticalDepth, sunDirection, dir, atmosphere_minus_shadow.xy, center, diffuse, uneclipsed);
+	skipRay(opticalDepth, dir, atmosphere_minus_shadow.yz, center);
+	processRay(sumR, sumM, opticalDepth, sunDirection, dir, atmosphere_minus_shadow.zw, center, diffuse, uneclipsed);
+
+	float mu = dot(dir, sunDirection); // mu in the paper which is the cosine of the angle between the sun direction and the ray direction
+	float phaseR = 3.f / (16.f * 3.141592) * (1 + mu * mu);
+	float g = 0.76f;
+	float phaseM = 3.f / (8.f * 3.141592) * ((1.f - g * g) * (1.f + mu * mu)) / ((2.f + g * g) * pow(1.f + g * g - 2.f * g * mu, 1.5f));
```

Comment:
If this uses the same Mie phase function as the planet ring PR, please convert this to a call to that function once it is merged.